HllDruidCombine invalid payloadstart position

Hi,

I'm using the Druid HLL java implementation to compute HLL values in a Spark job and I'm inserting that output into a `VARBINARY(1031)` column. (https://github.com/apache/incubator-druid/blob/master/hll/src/main/java/org/apache/druid/hll/HyperLogLogCollector.java)

My insert uses a vertica COPY command with `FORMAT 'HEX'`, and I'm serializing it as follows:

`javax.xml.bind.DatatypeConverter.printHexBinary(hyperLogLogCollector.toByteArray())`

Then, when I use `HllDruidCombine` in a select statement I'm getting the following error:

`Caused by: java.sql.SQLNonTransientException: [Vertica][VJDBC](5861) ERROR: Error calling aggregateArrs() in User Function HllDruidCombine at [/usr/local/vertica-hyperloglog/SOURCES/src/HllDruidCombine.cpp:73], error code: 0, message: Exception while aggregating intermediates: [invalid payloadstart position:65529]`

Are there any known incompatibilities between the two implementations?
Or, am I perhaps doing something wrong?

Any help or hints would be appreciated!

Thanks!

--- a/SOURCES/src/HllDruidCombine.cpp
+++ b/SOURCES/src/HllDruidCombine.cpp
@@ -60,7 +60,7 @@ public:
     {
       HllDruid hll = HllDruid::wrapRawBuffer(
           reinterpret_cast<uint8_t *>(aggs.getStringRef(0).data()),
-          aggs.getTypeMetaData().getColumnType(0).getStringLength());
+          aggs.getStringRef(0).length());
       do
       {
         hll.fold(
@@ -70,7 +70,7 @@ public:
     }
     catch (std::exception &e)
     {
-      vt_report_error(0, "Exception while aggregating intermediates: [%s]", e.what());
+      vt_report_error(0, "Exception while aggregating intermediates: [%d]", e.what());
     }
   }
 
@@ -82,7 +82,7 @@ public:
     {
       HllDruid hll = HllDruid::wrapRawBuffer(
           reinterpret_cast<uint8_t *>(aggs.getStringRef(0).data()),
-          aggs.getTypeMetaData().getColumnType(0).getStringLength());
+          aggs.getStringRef(0).length());
       do
       {
         hll.fold(

--- a/SOURCES/src/HllDruidDistinctCount.cpp
+++ b/SOURCES/src/HllDruidDistinctCount.cpp
@@ -51,11 +51,10 @@ public:
     try
     {
       HllDruid hll = HllDruid::wrapRawBuffer(
-          reinterpret_cast<std::uint8_t *>(aggs.getStringRef(0).data()),
-          aggs.getTypeMetaData().getColumnType(0).getStringLength());
+          reinterpret_cast<uint8_t *>(aggs.getStringRef(0).data()),
+          aggs.getStringRef(0).length());
       do
       {
-
         hll.fold(
             reinterpret_cast<const uint8_t *>(argReader.getStringRef(0).data()),
             argReader.getStringRef(0).length());
@@ -74,8 +73,8 @@ public:
     try
     {
       HllDruid hll = HllDruid::wrapRawBuffer(
-          reinterpret_cast<std::uint8_t *>(aggs.getStringRef(0).data()),
-          aggs.getTypeMetaData().getColumnType(0).getStringLength());
+          reinterpret_cast<uint8_t *>(aggs.getStringRef(0).data()),
+          aggs.getStringRef(0).length());
       do
       {
         hll.fold(
@@ -88,6 +87,7 @@ public:
       vt_report_error(0, "Exception while combining intermediates: [%s]", e.what());
     }
   }
+
   virtual void terminate(ServerInterface &srvInterface,
                          BlockWriter &resWriter,
                          IntermediateAggs &aggs)

--- a/SOURCES/src/hll_druid.cpp
+++ b/SOURCES/src/hll_druid.cpp
@@ -201,12 +201,12 @@ void HllDruid::serialize(uint8_t *outBuffer, size_t &outLength)
   { // store sparsely;
     memcpy(outBuffer, payload, sizeof(Header));
     outLength = NUM_HEADER_BYTES;
-    for (int i = 0; i < NUM_BYTES_FOR_BUCKETS; ++i)
+    for (uint16_t i = 0; i < NUM_BYTES_FOR_BUCKETS; ++i)
     {
       if (payload[i + NUM_HEADER_BYTES] != 0)
       {
-        *(outBuffer + outLength + 0) = (i >> 8);
-        *(outBuffer + outLength + 1) = (i & 0x00ff);
+        *(outBuffer + outLength + 0) = ((i + NUM_HEADER_BYTES) >> 8);
+        *(outBuffer + outLength + 1) = ((i + NUM_HEADER_BYTES) & 0x00ff);
         *(outBuffer + outLength + 2) = payload[i + NUM_HEADER_BYTES];
         outLength += 3;
       }

